n3fit/src/n3fit/model_gen.py

```diff
@@ -544,135 +543,206 @@ def pdfNN_layer_generator(
+    # Adding of logs happens inside, but before the NN and adds a dimension there
+    nn_input_dimensions = 1 if use_feature_scaling else 2
+
+    # Define the main input
+    do_nothing = lambda x: x
+    if use_feature_scaling:
+        pdf_input = Input(shape=(None, pdf_input_dimensions), batch_size=1, name='scaledx_x')
+        process_input = do_nothing
+        extract_nn_input = Lambda(lambda x: op.op_gather_keep_dims(x, 0, axis=-1), name='x_scaled')
+        extract_original = Lambda(lambda x: op.op_gather_keep_dims(x, 1, axis=-1), name='x')
+    else:  # add log(x)
+        pdf_input = Input(shape=(None, pdf_input_dimensions), batch_size=1, name='x')
+        process_input = Lambda(lambda x: op.concatenate([x, op.op_log(x)], axis=-1), name='x_logx')
+        extract_original = do_nothing
+        extract_nn_input = do_nothing
+
+    model_input = {"pdf_input": pdf_input}
+
     if subtract_one:
-        layer_x_eq_1 = op.numpy_to_input(np.array(input_x_eq_1).reshape(1, 1))
+        input_x_eq_1 = [1.0]
+        if use_feature_scaling:
+            input_x_eq_1 = scaler(input_x_eq_1)[0]
+        # the layer that subtracts 1 from the NN output
+        subtract_one_layer = Lambda(op.op_subtract, name='subtract_one')
+        layer_x_eq_1 = op.numpy_to_input(
+            np.array(input_x_eq_1).reshape(1, 1), name='x_ones', custom_shape=(None, 1)
+        )  # Just to make shapes consistent
         model_input["layer_x_eq_1"] = layer_x_eq_1
 
-    # Evolution layer
-    layer_evln = FkRotation(input_shape=(last_layer_nodes,), output_dim=out)
+    # the layer that multiplies the NN output by the preprocessing factor
+    apply_preprocessing_factor = Lambda(op.op_multiply, name='prefactor_times_NN')
 
     # Photon layer
-    layer_photon = AddPhoton(photons=photons)
+    layer_photon = AddPhoton(photons=photons, name="add_photon")
 
     # Basis rotation
-    basis_rotation = FlavourToEvolution(flav_info=flav_info, fitbasis=fitbasis)
+    basis_rotation = FlavourToEvolution(
+        flav_info=flav_info, fitbasis=fitbasis, name="pdf_evolution_basis"
+    )
+
+    # Evolution layer
+    layer_evln = FkRotation(input_shape=(last_layer_nodes,), output_dim=out, name="pdf_FK_basis")
 
     # Normalization and sum rules
     if impose_sumrule:
-        sumrule_layer, integrator_input = msr_impose(
+        sumrule_layer, integrator_input = generate_msr_model_and_grid(
             mode=impose_sumrule, scaler=scaler, photons=photons
         )
         model_input["integrator_input"] = integrator_input
     else:
         sumrule_layer = lambda x: x
+    # The photon is treated separately, need to get its integrals to normalize the pdf
+    if photons:
+        photon_integrals = photons.integral
+    else:
+        photon_integrals = [0.0 for _ in range(len(seed))]
+    import tensorflow as tf
+    photon_integrals = [tf.constant(photon_integrals[i_replica])
+                            for i_replica in range(len(seed))]
```

Comment:
Please move this down to where it's actually passed to the `sumrules_layer` to keep relevant bits of code together. 

Though I don't really have a preference regarding taking the specific replica inside the `Photon` class or not (this is something you can discuss with @scarlehoff), I don't really like this solution and would rather see the if-statement inside the loop where it's passed to the `sumrules_layer`.

Reply:
Oh first of all I didn't mean to leave the import statement there ofcourse, sorry about that, I can just use the existing `op.numpy_to_tensor`.
My aim with putting it there was to first define everything and then keep the final loop as clean as possible. But no matter what you'll have some somewhat confusing code somewhere due to the photon. I applied your suggestion, I agree it's a bit cleaner.